### ISSUE TITLE
Notification log design

### ISSUE THREAD

**Issue by buybackoff** [DataSpreads/DataSpreads#3]
NotificationLog, often called Log0 in code/comments, acts like a message bus to notify all processes about new messages in stream logs (SLs). SLs are independent and one could spin on them individually, but they are designed to be very lightweight and granular so that it's better to have 100s of logs and merge them if needed (with Spreads Zip) than to have a fat log and filter it.

Both are somewhat modeled after Aeron LogBuffer, but SL has quite different implementation. Log0 follows the original idea of atomically incrementing global position and thus reserving space for writes. But message size in Log0 is always just 8 bytes.

**True sharing vs false sharing**
Incrementing a shared long is the global bottleneck. Due to physical limitations of CPU we cannot do more than c.35M increments in a highly contended case (on i7-8700). This is `true sharing` (vs false sharing) and we cannot do anything. However 35 Mops is more than enough to start with. There is an easy solution to scale this - just add several notification channels depending on SL id, e.g. `(slid & 3)`. Then when Log0 reader spins it could just round-robin over the channels.

Another problem is we do have false sharing when writing 8 bytes payload values to the same cache line from different threads. Aeron does not have this problem because a message is at least 64 bytes (32 bytes header + 32 bytes alignment). Splitting into channels will help here as well.

We will lose total order by round-robin probing of the channels: different readers will have different sequences. However, we do not depend on the total order of notifications. Instead we have TimeService that guarantees total order of timestamps or we assume that an external data source has a sequence number or TS. Then we could open SLs `AsTimeSeries()` and Zip on timestamps. This is the intended use case.

**Thread preemption problem**
Due to the lightweight nature and the high count of streams we have a problem caused by thread preemption:
1. A writer thread `X` increments the counter to a version `V`;
2. It is being preempted before it writes the payload in the reserved slot;
3. Other writers increment & write payload beyond the slot from 1.
4. Reader is blocked on the slot from 1.
5. Some time later thread `X` is scheduled back and writes the payload, but:
a. Block that contains version `V` could be no longer active, thread `X` needs to find the block for the version `V`.
b. The reader was blocked on the version `V` and did not woke up listeners for all other notifications past the version `V`. Kind of head of line blocking.
c. The reader keeps reference to the block and it is not reclaimed by Packer.
d. The reader has no way to know if the writer `X` will come back or died (it could have been a separate process).

* `5.a` could be fixed by looking up a block from index as needed - slow, but the case is quite rare.
* `5.b` is a real problem for latency and somewhat for throughput (after unblocking Log0 reader will read accumulated values very fast, but this leads to a spike in overall load to process messages that should have been processed in real time)
* `5.c` we could live with that, reclaimed blocks could be reused by SLs and the space required is not too big given 8 bytes payload, unless `5.d`
* `5.d` if a writer dies after incrementing the global version but before writing the payload there is no way in the initial implementation to make progress.

So `5.b` and `5.d` *require* that we have a mechanism to skip slots from preempted writers.

**Requirements**

* High throughput. Limited by CPU bus.
* Minimal latency. `5.b` is the problem.
* Never miss a notification. A single missed notification could paralyze the entire system. We need unlocking mechanism and a dumb background task that periodically tries to wake up waiters.
* WAL could start from where it left. WAL was not included because it was a very early POC. What it needs is a binary tree of dirty SLs. Whenever a SL is opened it's current WAL position is written to the tree as `WALPosition -> Slid`. Since the tree is sorted, we know what Slids are not making a progress and we could flush them to storage (which even now supports writing incomplete last block, but this is turned off). `WALPosition` is by definition unique (it's `long` offset from WAL zero byte, 2^64 bytes is enough). So WAL's job is not to catch up with Log0, but to monitor dirty SLs. WAL must not depend on Log0, which version should be ephemeral.

**Non-requirements**
* Exactly-once notification. Payload has Slid and event type. Then execution goes on thread pool. If AsyncCursor could move then continuation is executed on TP, otherwise AC continues to wait. So multiple notifications are not an issue if they are limited in number and do not create fake load on TP.
* Reader order matches actual writes to SLs. We actually cannot achieve this. Writes happen for waiting readers only when Log0 is updated with payload.

**Unblocking**

Keep track of skipped slots and retry instead of yielding.

Reader logic:
1. Spin until next value. Before each/Nth pause (`SpinWait.SpinOnce`) read global version. If it is above the current position try to read the next slot(s). If there are no skipped values then try to ack writers who requested that and then yield/sleep. Optionally use local host UDP multicast as a blocking notification channel. If there are skipped slots, go to 3.
2. If the next slot is not empty, add current slot to an in-memory skipped queue and continue reading.
3. Take the oldest slot from the skipped queue, try to read it. MoveNext if successful, otherwise add the slot to the end of the queue unless 4.
4. If the slot belongs to a block is older then the previous block then drop the slot. Fire a task that opportunistically tries to unblock all open SLs. This requires some coordination with writers.

Writer logic:
1. Store active block pointer at the very beginning of Append.
2. Increment global version and try to write payload.
3. If the version slot is not in the active block then rotate the active block.
4. If the version now is in the active or previous block then write it.
5. [*this was missing before, but requires one additional read on every Append*] After writing payload read the current global version. If the written slot is not in the global active block or global previous block then just write the same payload as if recursively calling Append (but avoid actual recursion).

2-4 are problematic without 5. A process could be very inactive (and with low priority) and it could have a single stream. The active block could match the version, but then the writer thread is suspended. After the thread resumes it doesn't see a need for rotation and just writes the payload to the version slot. But other processes have run away and readers dropped the block to where the resumed thread is writing.

But different problems: suspended thread still keeps a reference to the block, we cannot reclaim it, or if we do the resumed thread could write to a reused block.

TBC...
